keycloak-claims-mapper/src/main/java/ee/eki/tolkevarav/sso/keycloakserviceprovider/tokenclaimsmapper/ClaimsFromTolkevaravApiTokenEnricher.java

```
@@ -0,0 +1,189 @@
+import java.time.Duration;
+
+import static java.net.http.HttpResponse.BodyHandlers.ofString;
+
+class ClaimsFromTolkevaravApiTokenEnricher {
+
+    private static final String SELECTED_INSTITUTION_ID_HEADER_KEY = "X-Selected-Institution-ID";
+    private static final String TARA_SUBJECT_CLAIM_ATTRIBUTE_KEY = "tara_subject_claim";
+    private final ConfigurationParameters configuration;
+    private final KeycloakSession keycloakSession;
+    private final UserSessionModel userSession;
+    private final ClientSessionContext clientSessionContext;
+
+    private static final Logger logger = Logger.getLogger(ClaimsFromTolkevaravApiTokenEnricher.class);
+
+    ClaimsFromTolkevaravApiTokenEnricher(ProtocolMapperModel mapperModel,
+                                                KeycloakSession keycloakSession,
+                                                UserSessionModel userSession,
+                                                ClientSessionContext clientSessionContext) {
+        this.configuration = ConfigurationParameters.fromModel(mapperModel);
+        this.keycloakSession = keycloakSession;
+        this.userSession = userSession;
+        this.clientSessionContext = clientSessionContext;
+    }
+
+    void enrichToken(IDToken token) {
+        try {
+            if (configuration.isInvalid()) {
+                throw new AcceptableMapperException("Mapper configuration is invalid: " + configuration);
+            }
+
+            var claims = queryClaimsFromApi(
+                retrievePersonalIdentificationCode(),
+                retrieveInstitutionId(),
+                generateAccessToken()
+            );
+
+            logger.infof("Claims received from Tõlkevärav API: %s. Adding it to claims in token.", claims);
+            token.getOtherClaims().put("tolkevarav", claims);
+        } catch (AcceptableMapperException exception) {
+            logger.info("Cancelled enriching of token because of an acceptable scenario: %s.", exception);
```

It seems that if request fails to authorization service, then this exception is handled, but token is still returned for the user. I think this will rise some issues where we cant clearly distinguish the error, was it that the token didnt contain enough information or did was the issue somewhere else. We mostly rely on the authorization service information in every use case where we need to consume the token then I would make it a requirement that no token is returned if info from authorization service can't be retrieved.

This can be made into separate configuration parameter, which helps to toggle this restriction and is helpful when developing. (e.g. if auth service is down then FE devs cant login to develop FE functionalities that don't require API

---

That is a very good point! The front-end relies on the `tolkevarav` claims being populated and if we "silently" leave it empty here, then it might lead to some hard-to-track-down issues later. I will add a configuration parameter for whether `UnacceptableMapperException`s are swallowed or cause failure.

---

Fixed with revision [ec3efd0](https://github.com/keeleinstituut/tv-sso/pull/5/commits/ec3efd0a125e6a62e31e004251d885b30e8feae3)